### PR TITLE
secrets: improve test coverages by testing fail connections

### PR DESCRIPTION
**Pull request by shantuo**
Updates: #1162

### CODE REVIEW

**Comment by shantuo** [comment]
The only exception is localsecrets (84%), which has two lines of return err's not tested. I think we might be ok with those?

File: secrets/awskms/kms_test.go [hunk 1]
```diff
@@ -57,3 +55,27 @@ func newHarness(ctx context.Context, t *testing.T) (drivertest.Harness, error) {
 func TestConformance(t *testing.T) {
 	drivertest.RunConformanceTests(t, newHarness)
 }
+
+// KMS-specific tests.
+
+func TestNoConnectionError(t *testing.T) {
+	if _, err := Dial(nil); err == nil {
```
Comment (vangent) [inline secrets/awskms/kms_test.go]: Nit: this could be a separate test.

File: secrets/awskms/kms_test.go [hunk 2]
```diff
@@ -57,3 +55,27 @@ func newHarness(ctx context.Context, t *testing.T) (drivertest.Harness, error) {
...
+
+	os.Setenv("AWS_ACCESS_KEY", "myaccesskey")
+	os.Setenv("AWS_SECRET_KEY", "mysecretkey")
+	os.Setenv("AWS_REGION", "us-east-1")
+	sess, err := session.NewSession()
+	if err != nil {
+		t.Fatal(err)
+	}
+	ctx := context.Background()
+	client, err := Dial(sess)
```
Comment (vangent) [inline secrets/awskms/kms_test.go]: Should check err here.

File: secrets/awskms/kms_test.go [hunk 3]
```diff
@@ -57,3 +55,27 @@ func newHarness(ctx context.Context, t *testing.T) (drivertest.Harness, error) {
...
+	}
+
+	os.Setenv("AWS_ACCESS_KEY", "myaccesskey")
+	os.Setenv("AWS_SECRET_KEY", "mysecretkey")
+	os.Setenv("AWS_REGION", "us-east-1")
+	sess, err := session.NewSession()
+	if err != nil {
+		t.Fatal(err)
+	}
+	ctx := context.Background()
```
Comment (vangent) [inline secrets/awskms/kms_test.go]: Nit: `ctx` and `plaintext` are only used once, so just use them inline, it's easier to read.

File: secrets/gcpkms/kms_test.go [hunk 4]
```diff
@@ -76,3 +74,27 @@ func newHarness(ctx context.Context, t *testing.T) (drivertest.Harness, error) {
 func TestConformance(t *testing.T) {
 	drivertest.RunConformanceTests(t, newHarness)
 }
+
+// KMS-specific tests.
+
```
Comment (vangent) [inline secrets/gcpkms/kms_test.go]: Same comments here as above.

File: secrets/vault/vault_test.go [hunk 5]
```diff
@@ -89,3 +82,32 @@ func testVaultServer(t *testing.T) (*api.Client, func()) {
 func TestConformance(t *testing.T) {
 	drivertest.RunConformanceTests(t, newHarness)
 }
+
+// Vault-specific tests.
+
+func TestNoConnectionError(t *testing.T) {
```
Comment (vangent) [inline secrets/vault/vault_test.go]: Same as above.

**Comment by vangent** [comment]
> The only exception is localsecrets (84%), which has two lines of return err's not tested. I think we might be ok with those?

The second one looks like it could be covered by a conformance test: try to decrypt something that wasn't encrypted with the key; it should fail. This can be a separate PR though.

File: secrets/awskms/kms_test.go (hunk 1 above)
Comment (shantuo) [reply secrets/awskms/kms_test.go]: Done.

File: secrets/awskms/kms_test.go (hunk 3 above)
Comment (shantuo) [reply secrets/awskms/kms_test.go]: Done.

File: secrets/awskms/kms_test.go (hunk 2 above)
Comment (shantuo) [reply secrets/awskms/kms_test.go]: Done.

File: secrets/gcpkms/kms_test.go (hunk 4 above)
Comment (shantuo) [reply secrets/gcpkms/kms_test.go]: Done.

File: secrets/vault/vault_test.go (hunk 5 above)
Comment (shantuo) [reply secrets/vault/vault_test.go]: Done.

**Comment by shantuo** [comment]
I need to update some tests after jba's PR.

**Comment by shantuo** [comment]
> > The only exception is localsecrets (84%), which has two lines of return err's not tested. I think we might be ok with those?
> 
> The second one looks like it could be covered by a conformance test: try to decrypt something that wasn't encrypted with the key; it should fail. This can be a separate PR though.

Except AWS's decryption relies on the key information inside the encryption body, and they don't take into a key in decryption. This is not conformance across all providers.

**Comment by vangent** [comment]
> Except AWS's decryption relies on the key information inside the encryption body, and they don't take into a key in decryption. This is not conformance across all providers.

Per discussion, let's add a conformance test that calls `Decrypt` with some random bytes (e.g., `[]byte("hello world")`). It should fail on all providers, even AWS.

**Pull request merged**